palindrome_number.py

```diff
@@ -0,0 +1,89 @@
+# https://leetcode.com/problems/palindrome-number/
+
+# Given an integer x, return true if x is palindrome integer.
+
+# An integer is a palindrome when it reads the same backward as forward.
+
+# For example, 121 is a palindrome while 123 is not.
+
+# Example 1:
+
+# Input: x = 121
+# Output: true
+# Explanation: 121 reads as 121 from left to right and from right to left.
+# Example 2:
+
+# Input: x = -121
+# Output: false
+# Explanation: From left to right, it reads -121. From right to left, it becomes 121-. Therefore it is not a palindrome.
+# Example 3:
+
+# Input: x = 10
+# Output: false
+# Explanation: Reads 01 from right to left. Therefore it is not a palindrome.
+ 
+
+# Constraints:
+
+# -231 <= x <= 231 - 1
+
+# Follow up: Could you solve it without converting the integer to a string?
+
+###
+
+test_1 = 121
+test_2 = -121
+test_3 = 123
+test_4 = -2
+
+def string_solution(x): 
+    is_palindrome = False
+
+    s = str(x)
+    s_backwards = s[::-1]
+
+    if s == s_backwards: 
+        is_palindrome = True
+
+    return is_palindrome
+
+
+#print(string_solution(test_1), string_solution(test_2), string_solution(test_3), string_solution(test_4))
+
+#Runtime: 76 ms, faster than 55.30% of Python online submissions for Palindrome Number.
+#Memory Usage: 13.5 MB, less than 40.03% of Python online submissions for Palindrome Number.
+
+# How to solve without turning into string? 
+
+import math
+
+def integer_solution(x): 
+    is_palindrome = False
+    no_digits = 0
+
+    if x > 0: 
+        no_digits = int(math.log10(x)+1)
+    elif x == 0:
+        no_digits = 1
+    else: 
+        no_digits = int(math.log10(-x)+2)
+
+    remaining_number = x
+    digit_list = []
+
+    for i in range(no_digits-1,-1,-1): 
+        digit = remaining_number//(10**(i))
+        digit_list.append(digit)
+        remaining_number = remaining_number - digit*(10**(i))
+
+    backwards_digit_list = digit_list[::-1]
+
+    if digit_list == backwards_digit_list: 
+        is_palindrome = True
+
+    return is_palindrome
+
+print(integer_solution(test_1), integer_solution(test_2), integer_solution(test_3), integer_solution(test_4))
+
+# Runtime: 82 ms, faster than 49.00% of Python online submissions for Palindrome Number.
+# Memory Usage: 13.4 MB, less than 66.69% of Python online submissions for Palindrome Number.
```

Comment:
Hmm ... this solution is pretty complex. Notice it's a bit slower than the first one. This is largely due to the extra effort required to convert the number to an array and then reverse and compare the arrays.

An ideal integer (only) solution uses some math trick to reverse the integer. The trick here will be to use the [modulo operator](https://www.askpython.com/python/python-modulo-operator-math-fmod), `%`. _The gist of the operator is that it returns the remainder of a division operation._

Using modulo and some extra operators, we can have a reverse function like so:
```python
def reverse(num):
    reversed = 0
    while num > 0:
      reversed = (reversed * 10) + (num % 10)
      num = num // 10

    return reversed
```
(Try it: https://trinket.io/python/31bf271606)

Now reversed, you can complete the rest of the algorithm.

Reply:
Thank you! We did several exercises with modulo in my course, but somehow I totally forgot about it! Would you advice to replace the log10 with modulo entirely or just write another solution to the challenge?

Reply:
Sure, you can add it as another solution to the challenge